Implement multiple paginators

josegonzalez opened the pull request:
This change provides the initial support for multiple paginators on a given page.

To use it, you can do something like the following in your controller layer:

``` php
$users = $this->paginate($this->Users->find(), ['prefix' => 'users']);
$categories = $this->paginate($this->Categories->find(), ['prefix' => 'categories']);

$this->set(compact('users', 'categories'));
```

The in your view layer, specify the model option whenever the `PaginatorHelper` allows you to set options.

Closes #1731

josegonzalez committed:
Implement mulitple paginators

This change provides the initial support for multiple paginators on a given page.

To use it, you can do something like the following in your controller layer:

```php
$users = $this->paginate($this->Users->find(), ['prefix' => 'users']);
$categories = $this->paginate($this->Categories->find(), ['prefix' => 'categories']);

$this->set(compact('users', 'categories'));
```

The in your view layer, specify the model option whenever the `PaginatorHelper` allows you to set options.

Closes #1731

thisispaulcyr commented:
Thanks for the work on this! I'm eagerly awaiting its completion. Are there any updates on the status?

markstory commented:
@thisispaulcyr These changes need tests written before they will be merged in. Would you be able/willing to help with that?

thisispaulcyr commented:
@markstory Yes, I would be interested. However, I have never done it before. Is it something that can be learned in a couple of days? If so, are you able to point me in the right direction to get started?

markstory commented:
The [book has a small section](http://book.cakephp.org/3.0/en/contributing/code.html) on getting started with contributing code.

josegonzalez committed:
Add tests for mergeOptions with custom prefix

josegonzalez committed:
Test Controller::paginate() with custom settings

josegonzalez committed:
Add tests for setting the default model

josegonzalez commented:
I added tests for a few things except the url generation, which I must admit I do not know quite *how* to test.

josegonzalez committed:
Remove extra newline

ADmad commented on src/View/Helper/PaginatorHelper.php:
```diff
@@ -412,10 +419,16 @@ public function sort($key, $title = null, array $options = [])
 
         $sortKey = $this->sortKey($options['model']);
         $defaultModel = $this->defaultModel();
+        $model = Hash::get($options, 'model', $defaultModel);
```
Call to `Hash::get()` is not necessary, can just do `$model = $options['model'] ?: $defaultModel;`

ADmad commented on src/Controller/Component/PaginatorComponent.php:
```diff
@@ -202,7 +203,8 @@ public function paginate($object, array $settings = [])
             'sortDefault' => $sortDefault,
-            'directionDefault' => $directionDefault
+            'directionDefault' => $directionDefault,
+            'prefix' => Hash::get($options, 'prefix', null),
```
Call to `Hash::get()` can be avoided by adding `'prefix' => null` to default options array on line 162.

markstory committed:
Incorporate feedback on pull request.

markstory committed:
Add URL generation tests for multiple pagination.

Add tests similar tests to what we have for routing prefixes.

markstory commented:
@josegonzalez I added some tests around URL generation in the helper.

markstory commented:
I'm a little concerned that `prefix` overlaps with routing prefixes in a confusing way. Some alternatives I thought of are:

* `queryStringPrefix`
* `parameterNamespace`
* `queryNamespace`
* `requestScope`

Do any of these (or another option) work better for people?

htstudios commented:
I like requestScope the most, as it could possibly be applied to POST, too (say PRG).

lorenzo commented:
`queryScope` ?

ADmad commented:
Or just single worded `scope`?

markstory commented:
@ADmad scope works too.

markstory committed:
Rename 'prefix' => 'scope'.

I am concerned that 'prefix' will be easily confused with routing
prefixes, as both this prefix and routing prefixes affect URL
generation. scope is a slightly less confusing term to me.

lorenzo commented:
Tests failing :(

ADmad commented:
The new scope option needs to be documented under relevant docblock.

markstory committed:
Fix controller tests.

markstory commented:
Tests and API docs updated. :blush:

htstudios commented:
Thank you guys!
This does mean that bake could now support multiple related pagination on view actions, right?

dereuromark commented:
:+100:
